Enable build on Ubuntu 22.04 runner

Nicky-D opened the pull request:
- Update config.guess/.sub to latest version
- Remove obsolete ltversion.m4
- Reconfigure via autoconf
- Remove std=c++17 from compile options (apr is a pure C project and having a C++ option will cause the build to fail due to -Werror)

Nicky-D committed:
Enable Ubuntu runner

Nicky-D committed:
Call autoreconf -vif

Nicky-D committed:
Update config.sub and config.guess

Nicky-D committed:
Try to work around ' configure: error: could not determine the string function for int64_t' on the ubuntu 22.04 runner

Nicky-D committed:
Remove ltversion.m4, this file is not supposed to be here.

Nicky-D committed:
For now remove -Werror for now, in the Ubuntu 22.04 VM it is hard to debug what really causes a warning.

Nicky-D committed:
Remove -std=c++17 from compile options, this is likely what causes the warnings and thus errors during configure.

Nicky-D committed:
Try to bring back -Werror.

Nicky-D commented:
@vir-linden / @nat-goodspeed this PR is sitting around without even much as a single peep for nearly a week now.
I need those builds as input for other PRs otherwise I've to roll my own under my GitHub account, which I'd like to avoid if possible. 

Is there any ETA when PRs are acted on?

nat-goodspeed commented on build-cmd.sh:
```diff
@@ -271,20 +271,19 @@ case "$AUTOBUILD_PLATFORM" in
 
     opts="-m$AUTOBUILD_ADDRSIZE $LL_BUILD_RELEASE"
-
+    opts=$(echo $opts | sed -e 's/-std=c++17//')
```
The `remove_cxxstd` shell function from [secondlife/build-variables](https://github.com/secondlife/build-variables/blob/master/functions#L79-L85) removes that switch regardless of the specific version.

Nicky-D replied:
Would this be expected to be called automagically or does it need to be called manually? I suppose the latter, as it did not kick in for me.
And I did not try to call it myself (I was not aware of this function). > Will try to switch over to it and then update the  PR accordingly

nat-goodspeed reviewed (commented):
Is this a new update from upstream? If so we would typically merge the upstream update to the `vendor` branch, then the updated `vendor` branch to the `master` branch.

But since the only changes that seem to affect upstream code are to `apr*/build/config.{guess,sub}`, with no version changes or anything, I'm guessing no. I'd still like confirmation though.

Nicky-D commented:
> Is this a new update from upstream? If so we would typically merge the upstream update to the `vendor` branch, then the updated `vendor` branch to the `master` branch.
> 
> But since the only changes that seem to affect upstream code are to `apr*/build/config.{guess,sub}`, with no version changes or anything, I'm guessing no. I'd still like confirmation though.

As far as I am aware are we caught up on APR (APR2 is in beta for ages now, but has no release. So no bueno).

Those config files are just the latest version from autotools, grabbed from here:
config.sub 'https://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.sub;hb=HEAD'
config.guess 'https://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess;hb=HEAD'

Nicky-D committed:
Replace call to sed with remove_cxxstd

As per Nat's suggestion. build_variables has a snazzy function for this. No need to mess around with sed

nat-goodspeed commented:
If the `config.{sub,guess}` files are the same for all three target libraries, I wonder if it would make sense to commit symlinks in `apr-util/build` and `apr-iconv/build`? Or even make the `build` subdirectory itself a symlink?

Nicky-D commented:
> If the `config.{sub,guess}` files are the same for all three target libraries, I wonder if it would make sense to commit symlinks in `apr-util/build` and `apr-iconv/build`? Or even make the `build` subdirectory itself a symlink?

1. Historically Windows has very weak symlink support. I'm not sure how well this would go there and how git handles it there.
2. I believe an apr vendor drop is handled by unzipping each release of apr, apr-iconv and apr-util into it's subdirectory. This again would interfere if there is a symlink
3. While config.sub/guess are identical, (some) file in the build directory differ

yaynstuff commented:
> Is this a new update from upstream? If so we would typically merge the upstream update to the `vendor` branch, then the updated `vendor` branch to the `master` branch.
> 
> But since the only changes that seem to affect upstream code are to `apr*/build/config.{guess,sub}`, with no version changes or anything, I'm guessing no. I'd still like confirmation though.

Have you guys made a decision on what version of ubuntu you are building on? I notice for webrtc you have picked the 20.04 LTS release, which is what we use to build the firestorm linux64 builds since we try to at least support all LTS releases, not just the newest one (gives us more coverage)

But in some cases you are using distros with a newer toolchain... Trying to link against your APR I see stuff like this:

```
/usr/bin/ld: /home/pork/src/viewer/current-pbr/build-linux-x86_64/packages/lib/release/libapr-1.so: undefined reference to `pthread_attr_setguardsize@GLIBC_2.34'
/usr/bin/ld: /home/pork/src/viewer/current-pbr/build-linux-x86_64/packages/lib/release/libapr-1.so: undefined reference to `pthread_mutexattr_setprotocol@GLIBC_2.34'
/usr/bin/ld: /home/pork/src/viewer/current-pbr/build-linux-x86_64/packages/lib/release/libapr-1.so: undefined reference to `pthread_mutexattr_setrobust@GLIBC_2.34'
/usr/bin/ld: /home/pork/src/viewer/current-pbr/build-linux-x86_64/packages/lib/release/libapr-1.so: undefined reference to `sem_trywait@GLIBC_2.34'
/usr/bin/ld: /home/pork/src/viewer/current-pbr/build-linux-x86_64/packages/lib/release/libapr-1.so: undefined reference to `dlclose@GLIBC_2.34'
/usr/bin/ld: /home/pork/src/viewer/current-pbr/build-linux-x86_64/packages/lib/release/libapr-1.so: undefined reference to `fstat@GLIBC_2.33'
```
On Ubuntu 20.04 LTS:
```
pork@focal64:~/src/viewer/current-pbr$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.6 LTS
Release:        20.04
Codename:       focal
pork@focal64:~/src/viewer/current-pbr$ dpkg -l | grep libc6
ii  libc6:amd64                            2.31-0ubuntu9.14                  amd64        GNU C Library: Shared libraries
ii  libc6-dev:amd64                        2.31-0ubuntu9.14                  amd64        GNU C Library: Development Libraries and Header Files
```
So what is the deal here, do we need to be building with 22.04 LTS in order to use LL's recent linux64 3p libraries? The glibc in 20.04 is obviously not new enough.

yaynstuff commented:
Its also evident you're not using Ubuntu 20.04 for building jsoncpp either since its obvious its been built with gcc11 which is not available on Ubuntu 20.04 LTS

nat-goodspeed commented:
> Have you guys made a decision on what version of ubuntu you are building on?

We don't build or test a Linux viewer ourselves. We do accept PRs that enable Linux builds of the autobuild packages and the viewer, though. If there's build.yaml magic that works better for you, please submit that PR.

yaynstuff commented:
> > Have you guys made a decision on what version of ubuntu you are building on?
> 
> We don't build or test a Linux viewer ourselves. We do accept PRs that enable Linux builds of the autobuild packages and the viewer, though. If there's build.yaml magic that works better for you, please submit that PR.

Hi Nat - I think it would be good to come to an agreement on what linux distro to build your libraries on, and make that a default for all of your linux64 builds -- that way any TPV projects that want to consume them know what to expect. But to have one library built with one version of ubuntu, and then another library built with a newer version will lead to confusion.

Are most of them being build on 22.04?

We're happy to adopt this for firestorm if thats the case. I've been maintaining Firestorm's own 3p libraries for years and am kind of getting tired of it - I'd like to start consuming the LL builds (just like we currently do for mac and windows).

nat-goodspeed commented:
In the `build.yaml` for all our `3p-*` repos I checked, the `jobs.build.strategy.matrix.os` line specifically cites `ubuntu-22.04`. There are some common-platform packages that run on `ubuntu-latest` instead, but presumably that's only used for source-only packages.

yaynstuff commented:
Ok, we can work with that, I will talk to Beq about updating our build runners to use 22.04. Thank you :)
